src/Simplex_tree/include/gudhi/Simplex_tree/hooks_simplex_base.h

```diff
@@ -0,0 +1,69 @@
+/*    This file is part of the Gudhi Library - https://gudhi.inria.fr/ - which is released under MIT.
+ *    See file LICENSE or go to https://gudhi.inria.fr/licensing/ for full license details.
+ *    Author(s):       Clément Maria
+ *
+ *    Copyright (C) 2020 Inria
+ *
+ *    Modification(s):
+ *      - YYYY/MM Author: Description of the modification
+ */
+
+#ifndef SIMPLEX_TREE_HOOKS_SIMPLEX_BASE_H_
+#define SIMPLEX_TREE_HOOKS_SIMPLEX_BASE_H_
+
+#include <boost/intrusive/list.hpp>
+
+namespace Gudhi {
+  /** \brief Data structure to put all simplex tree nodes with same label into a list.
+   *
+   * Allows one to access all subtrees of the simplex tree rooted at a node with a given label.
+   * Used in particular for fast cofaces location, and fast insertion and deletion of edges in a flag complex.
+   *
+   * Only if SimplexTreeOptions::link_nodes_by_label is true, otherwise store nothing.
+   */
+
+  // no hook
+  struct Hooks_simplex_base_dummy {};
+  // todo on Hooks_simplex_base_link_nodes:
+  // make the class movable but not copiable
+  // translate the boost macros into C++11 syntax (boost independent)
+  // update the Node concept and the doc
+  struct Hooks_simplex_base_link_nodes {
+   private:
+    BOOST_COPYABLE_AND_MOVABLE(Hooks_simplex_base_link_nodes)
+   public:
+    Hooks_simplex_base_link_nodes() {}
+    // the copy constructor, inherited by the Node class, exchanges hooks,
+    // and make the ones of this invalid.
+    // this is used when stored in a map like DS, using copies when
+    // performing insertions and rebalancing of the rbtree
+    Hooks_simplex_base_link_nodes(const Hooks_simplex_base_link_nodes& other) {
```

Comment:
I replaced boost syntax from Hooks_simplex_base_link_nodes with a c++11 syntax on 95475ea94 , but only a move syntax is not sufficient enough for Simplex_tree constructor and move unitary tests

Reply:
I fixed documentation and remove comments about movable/copyable part on bdbab2d61

Reply:
```diff
--- a/src/Simplex_tree/test/simplex_tree_unit_test.cpp
+++ b/src/Simplex_tree/test/simplex_tree_unit_test.cpp
@@ -593,6 +593,7 @@ BOOST_AUTO_TEST_CASE_TEMPLATE(NSimplexAndSubfaces_tree_insertion, typeST, list_o
 template<class typeST, class Vertex_handle>
 void test_cofaces(typeST& st, const std::vector<Vertex_handle>& expected, int dim, const std::vector<typename typeST::Simplex_handle>& res) {
   std::size_t nb_res = 0;
+  typeST marc = st;
   if (dim == 0) {
     typename typeST::Cofaces_simplex_range stars = st.star_simplex_range(st.find(expected));
     for (auto simplex = stars.begin(); simplex != stars.end(); ++simplex) {
```
causes the test to fail.

Reply:
You are right. I modified copy constructor to copy hooks instead of exchanging them on [b7322d5](https://github.com/GUDHI/gudhi-devel/pull/877/commits/b7322d5435e6f10cd2a236319b33a4da5f402f0e)

Reply:
Now this version fails:
```diff
--- a/src/Simplex_tree/test/simplex_tree_unit_test.cpp
+++ b/src/Simplex_tree/test/simplex_tree_unit_test.cpp
@@ -591,8 +591,9 @@ BOOST_AUTO_TEST_CASE_TEMPLATE(NSimplexAndSubfaces_tree_insertion, typeST, list_o
 }
 
 template<class typeST, class Vertex_handle>
-void test_cofaces(typeST& st, const std::vector<Vertex_handle>& expected, int dim, const std::vector<typename typeST::Simplex_handle>& res) {
+void test_cofaces(typeST& st_, const std::vector<Vertex_handle>& expected, int dim, const std::vector<typename typeST::Simplex_handle>& res) {
   std::size_t nb_res = 0;
+  typeST st = st_;
   if (dim == 0) {
     typename typeST::Cofaces_simplex_range stars = st.star_simplex_range(st.find(expected));
     for (auto simplex = stars.begin(); simplex != stars.end(); ++simplex) {
```

Reply:
Ok. I added some specific code and tests on [95ce710](https://github.com/GUDHI/gudhi-devel/pull/877/commits/95ce710cfbc2697de328b943e35b15becaf0e309) for copy/move constructor/assignement

Reply:
Maybe we could make things a bit less misleading now?
```diff
--- a/src/Simplex_tree/include/gudhi/Simplex_tree/hooks_simplex_base.h
+++ b/src/Simplex_tree/include/gudhi/Simplex_tree/hooks_simplex_base.h
@@ -28,12 +28,15 @@ namespace Gudhi {
   struct Hooks_simplex_base_link_nodes {
    public:
     Hooks_simplex_base_link_nodes() {}
+    // The following 2 functions are conceptually wrong, but when copying a Simplex_tree it is more convenient to have them (and fix things asap after they are called).
     // copy constructor, inherited by the Node class, requires to copy hooks
-    Hooks_simplex_base_link_nodes(const Hooks_simplex_base_link_nodes& other)
-    : list_max_vertex_hook_(other.list_max_vertex_hook_) { }
+    Hooks_simplex_base_link_nodes(const Hooks_simplex_base_link_nodes&)
+    : list_max_vertex_hook_() { }
     // copy assignment
-    Hooks_simplex_base_link_nodes& operator=(const Hooks_simplex_base_link_nodes& other) {
-      list_max_vertex_hook_.swap_nodes(other.list_max_vertex_hook_);
+    Hooks_simplex_base_link_nodes& operator=(const Hooks_simplex_base_link_nodes&) {
+#ifdef GUDHI_DEBUG
+      Member_hook_t().swap_nodes(list_max_vertex_hook_);
+#endif
       return *this;
     }
     // move constructor
@@ -52,7 +55,7 @@ namespace Gudhi {
         boost::intrusive::link_mode<boost::intrusive::auto_unlink>>
         Member_hook_t;
 
-    mutable Member_hook_t list_max_vertex_hook_;
+    Member_hook_t list_max_vertex_hook_;
   };
 
 }  // namespace Gudhi
```

Reply:
I took your suggestion as is on [9c0611f](https://github.com/GUDHI/gudhi-devel/pull/877/commits/9c0611f38a2d851f773357b04ba3e2e75e3bc76d). I didn't get why we cannot say:
```cpp
Hooks_simplex_base_link_nodes& operator=(const Hooks_simplex_base_link_nodes&) = delete;
```
(I have some compilation error)
instead of throwing an exception.

Reply:
Something like: internally, copying the array delegates the work to a more general function like insert, which needs operator= when there are already elements in the array. So we need operator= to compile, but the code using it is dead in our case.